src/sync_command_response_client.rs

```diff
@@ -0,0 +1,79 @@
+use crate::{
+    bus::SharedMessageBus,
+    mempool::{MempoolCommand, MempoolResponse},
+    model::Transaction,
+};
+use anyhow::{Context, Result};
+use tracing::info;
+
+trait CmdRespClient<Cmd, Resp> {
+    async fn request(&self, cmd: Cmd) -> Result<Option<Resp>>;
+}
+
+impl CmdRespClient<MempoolCommand, MempoolResponse> for SharedMessageBus {
+    async fn request(&self, cmd: MempoolCommand) -> Result<Option<MempoolResponse>> {
+        info!("Sending command {:?}", cmd);
+        _ = self.sender::<MempoolCommand>().await.send(cmd.clone());
+        match cmd {
+            MempoolCommand::GetTxs => {
+                let mut receiver = self.receiver::<MempoolResponse>().await;
+
+                loop {
+                    if let MempoolResponse::Txs { txs } = receiver
+                        .recv()
+                        .await
+                        .context("Receiving mempool response")?
+                    {
+                        return Ok(Some(MempoolResponse::Txs { txs }));
+                    }
+                }
+            }
+            MempoolCommand::CreatePendingBatch { id: req_id } => {
+                let mut receiver = self.receiver::<MempoolResponse>().await;
+
+                loop {
+                    if let MempoolResponse::PendingBatch { id: resp_id, txs } = receiver
+                        .recv()
+                        .await
+                        .context("Receiving mempool response")?
+                    {
+                        if req_id == resp_id {
+                            return Ok(Some(MempoolResponse::PendingBatch { id: resp_id, txs }));
+                        }
+                    }
+                }
+            }
+            _ => Ok(None),
+        }
+    }
+}
+
+#[tokio::test]
+async fn test_bus() {
+    info!("Starting test");
+    let bus = SharedMessageBus::new();
+
+    let mut receiver = bus.receiver::<MempoolCommand>().await;
+    let sender = bus.sender::<MempoolResponse>().await;
+
+    tokio::spawn(async move {
+        loop {
+            info!("Received command");
+            if let Ok(MempoolCommand::GetTxs) = receiver.recv().await {
+                info!("Sending answer");
+                _ = sender.send(MempoolResponse::Txs {
+                    txs: vec![Transaction::default()],
+                });
+            }
+        }
+    });
+
+    let resp = bus.request(MempoolCommand::GetTxs).await;
+
+    assert_eq!(
+        resp.unwrap().unwrap(),
+        MempoolResponse::Txs {
+            txs: vec![Transaction::default(), Transaction::default()]
```

Comment:
why 2 txs in the expected response ?

Reply:
Sorry, did not see the message, funny thing, I don't know how to display logs when runnning tests, only failures are displayed, that is why I use failures

Reply:
just `export RUST_LOG=debug` before running test